File: Samples/LanguageConfigSample/ConfigForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace LanguageConfigSample
{
    public partial class ConfigForm : Form
    {
        Ezfx.Csv.CsvConfig config = new Ezfx.Csv.CsvConfig();

        public ConfigForm()
        {
            InitializeComponent();
            LoadEncoding();
            config = new Ezfx.Csv.CsvConfig(typeof(LanguagePopularity));
        }

        private void ConfigForm_Load(object sender, EventArgs e)
        {
            csvConfigBindingSource.DataSource = this.config;
            dataGridView1.DataSource = config.Columns;
            config.PropertyChanged += new PropertyChangedEventHandler(config_PropertyChanged);
        }

        void config_PropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            changedLabel.ForeColor = Color.Red;
            changedPropertyLabel.ForeColor = Color.Red;
            changedPropertyLabel.Text = e.PropertyName;
        }

        private void LoadEncoding()
        {
            config.CodePage = Encoding.Default.CodePage;
            List<EncodingInfo> infos = Encoding.GetEncodings().ToList();
            infos.Sort((a, b) => a.DisplayName.CompareTo(b.DisplayName));
            foreach (EncodingInfo info in infos)
            {
                encodingComboBox.Items.Add(info.DisplayName);
            }
        }
        
        private void dataGridView1_CellEndEdit(object sender, DataGridViewCellEventArgs e)
        {
            changedLabel.ForeColor = Color.Red;
            config.IsChanged = true;
        }

        private void cancelButton_Click(object sender, EventArgs e)
        {
            if(config.IsChanged)
            {
                if (DialogResult.OK == MessageBox.Show("Changes have been made, Save changes?", "", MessageBoxButtons.OKCancel))
                {
                    Save();
                }
            }
            this.Close();
        }

        private void okeyButton_Click(object sender, EventArgs e)
        {
            Save();
            this.Close();
        }

        private void Save()
        {
            MessageBox.Show("Save Not Implement.");
        }
    }
}

File: Samples.VB/LanguageConfigSampleVB/ConfigForm.vb
Imports System.Text
Imports Ezfx.Csv
Imports System.ComponentModel

Public Class ConfigForm

    Dim config As CsvConfig = Nothing
    Public Sub New()
        InitializeComponent()
        LoadEncoding()
        config = New CsvConfig(GetType(LanguagePopularity))
    End Sub

    Private Sub ConfigForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        CsvConfigBindingSource.DataSource = Me.config
        dataGridView1.DataSource = config.Columns
        AddHandler config.PropertyChanged, AddressOf config_PropertyChanged
        '   config.PropertyChanged += new PropertyChangedEventHandler(config_PropertyChanged);
    End Sub

    Private Sub LoadEncoding()
        Me.encodingComboBox.Items.Clear()
        For Each enc As EncodingInfo In Encoding.GetEncodings().OrderBy(Function(c) c.DisplayName)
            Me.encodingComboBox.Items.Add(enc.DisplayName)
        Next
    End Sub

    Private Sub config_PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
        changedLabel.ForeColor = Color.Red
        changedPropertyLabel.ForeColor = Color.Red
        changedPropertyLabel.Text = e.PropertyName
    End Sub


    Private Sub dataGridView1_CellEndEdit(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DataGridViewCellEventArgs) Handles dataGridView1.CellEndEdit
        changedLabel.ForeColor = Color.Red
        config.IsChanged = True
    End Sub

    Private Sub okeyButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles okeyButton.Click
        Save()
        Me.Close()
    End Sub

    Private Sub cancelButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cancelButton.Click
        If config.IsChanged Then
            Dim dialogResult As DialogResult = MessageBox.Show("Changes have been made, Save?", "", MessageBoxButtons.YesNoCancel, Nothing)
            If dialogResult = dialogResult.Yes Then
                Save()
            ElseIf dialogResult = dialogResult.No Then
                Me.Close()
            End If
        End If

    End Sub

    Private Sub Save()
        MessageBox.Show("Save method not implement.")
    End Sub

End Class
